client-flink/flink-1.20/src/main/java/org/apache/celeborn/plugin/flink/tiered/CelebornTierConsumerAgent.java

```diff
@@ -0,0 +1,552 @@
+import org.apache.celeborn.plugin.flink.buffer.ReceivedNoHeaderBufferPacker;
+import org.apache.celeborn.plugin.flink.readclient.FlinkShuffleClientImpl;
+
+public class CelebornTierConsumerAgent implements TierConsumerAgent {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CelebornTierConsumerAgent.class);
+
+  private final CelebornConf conf;
+
+  private final int gateIndex;
+
+  private final List<TieredStorageConsumerSpec> consumerSpecs;
+
+  private final List<TierShuffleDescriptor> shuffleDescriptors;
+
+  /**
+   * partitionId -> subPartitionId -> reader, note that subPartitions may share the same reader, as
+   * a single reader can consume multiple subPartitions to improvement performance.
+   */
+  private final Map<
+          TieredStoragePartitionId, Map<TieredStorageSubpartitionId, CelebornChannelBufferReader>>
+      bufferReaders;
+
+  /** Lock to protect {@link #receivedBuffers} and {@link #cause} and {@link #closed}, etc. */
+  private final Object lock = new Object();
+
+  /** Received buffers from remote shuffle worker. It's consumed by upper computing task. */
+  @GuardedBy("lock")
+  private final Map<TieredStoragePartitionId, Map<TieredStorageSubpartitionId, Queue<Buffer>>>
+      receivedBuffers;
+
+  @GuardedBy("lock")
+  private final Set<Tuple2<TieredStoragePartitionId, TieredStorageSubpartitionId>>
+      subPartitionsNeedNotifyAvailable;
+
+  @GuardedBy("lock")
+  private boolean hasStart = false;
+
+  @GuardedBy("lock")
+  private Throwable cause;
+
+  /** Whether this remote input gate has been closed or not. */
+  @GuardedBy("lock")
+  private boolean closed;
+
+  private FlinkShuffleClientImpl shuffleClient;
+
+  /**
+   * The notify target is flink inputGate, used in notify input gate which subPartition contain
+   * shuffle data that can to be read.
+   */
+  private AvailabilityNotifier availabilityNotifier;
+
+  private TieredStorageMemoryManager memoryManager;
+
+  public CelebornTierConsumerAgent(
+      CelebornConf conf,
+      List<TieredStorageConsumerSpec> tieredStorageConsumerSpecs,
+      List<TierShuffleDescriptor> shuffleDescriptors) {
+    checkArgument(!shuffleDescriptors.isEmpty(), "Wrong shuffle descriptors size.");
+    checkArgument(
+        tieredStorageConsumerSpecs.size() == shuffleDescriptors.size(),
+        "Wrong consumer spec size.");
+    this.conf = conf;
+    this.gateIndex = tieredStorageConsumerSpecs.get(0).getGateIndex();
+    this.consumerSpecs = tieredStorageConsumerSpecs;
+    this.shuffleDescriptors = shuffleDescriptors;
+    this.bufferReaders = new HashMap<>();
+    this.receivedBuffers = new HashMap<>();
+    this.subPartitionsNeedNotifyAvailable = new HashSet<>();
+    for (TierShuffleDescriptor shuffleDescriptor : shuffleDescriptors) {
+      if (!(shuffleDescriptor instanceof TierShuffleDescriptorImpl)) {
+        continue;
+      }
+      initShuffleClient((TierShuffleDescriptorImpl) shuffleDescriptor);
+      break;
+    }
+    initBufferReaders();
```

Comment:
If `shuffleClient` is null, does this need to initialize buffer readers?

Reply:
if `shuffleClient` is null, initializing buffer readers is unnecessary.
In the current implementation, the `shuffleClient` should not be null. I will add a check for `shuffleClient`.